h2o-py/h2o/explain/_explain.py

```diff
@@ -0,0 +1,1763 @@
+        for exp in exclude_explanations:
+            if exp not in possible_explanations:
+                raise RuntimeError("Unknown explanation \"{}\". Please use one of: {}".format(
+                    exp, possible_explanations))
+        explanations = [exp for exp in possible_explanations if exp not in exclude_explanations]
+    return explanations
+
+
+def _process_models_input(
+        models,  # type: Union[h2o.automl._base.H2OAutoMLBaseMixin, List, h2o.model.ModelBase]
+        test_frame,  # type: h2o.H2OFrame
+):
+    # type: (...) -> Tuple[bool, List, bool, bool, bool, List, List]
+    """
+    Helper function to get basic information about models/H2OAutoML.
+
+    :param models: H2OAutoML/List of models/H2O Model
+    :param test_frame: H2O Frame
+    :return: tuple (is_aml, models_to_show, classification, multinomial_classification,
+                    multiple_models, targets, tree_models_to_show)
+    """
+    is_aml = isinstance(models, h2o.automl._base.H2OAutoMLBaseMixin)
+    if is_aml:
+        models_to_show = [models.leader]
+        multiple_models = models.leaderboard.nrow > 1
+    elif isinstance(models, h2o.model.ModelBase):
+        models_to_show = [models]
+        multiple_models = False
+    else:
+        models_to_show = models
+        multiple_models = len(models) > 1
+    tree_models_to_show = _get_tree_models(models, 1 if is_aml else float("inf"))
+    y = _get_xy(models_to_show[0])[1]
+    classification = test_frame[y].isfactor()[0]
+    multinomial_classification = classification and test_frame[y].nlevels()[0] > 2
+    targets = [None]
+    if multinomial_classification:
+        targets = [[t] for t in test_frame[y].levels()[0]]
+    return is_aml, models_to_show, classification, multinomial_classification, \
+           multiple_models, targets, tree_models_to_show
+
+
+def _custom_args(user_specified, **kwargs):
+    # type: (Optional[Dict], **Any) -> Dict
+    """
+    Helper function to make customization of arguments easier.
+
+    :param user_specified: dictionary of user specified overrides or None
+    :param kwargs: default values, such as, `top_n=5`
+    :return: dictionary of actual arguments to use
+    """
+    if user_specified is None:
+        user_specified = dict()
+
+    result = dict(**kwargs)
+    result.update(user_specified)
+
+    return result
+
+
+def explain(
+        models,  # type: Union[h2o.automl._base.H2OAutoMLBaseMixin, List[h2o.model.ModelBase]]
+        test_frame,  # type: h2o.H2OFrame
+        columns_of_interest=None,  # type: Optional[List[str]]
+        top_n_features=5,  # type: int
+        include_explanations="ALL",  # type: Union[str, List[str]]
+        exclude_explanations=[],  # type: Union[str, List[str]]
+        plot_overrides=dict(),  # type: Dict
+        figsize=(16, 9),  # type: Tuple[float]
+        render=True,  # type: bool
+        qualitative_colormap="Dark2",  # type: str
+        sequential_colormap="RdYlBu_r"  # type: str
+):
```

Comment:
- `columns_of_interest`: isn't it a bit verbose? why not just `columns`?
- `top_n_features`: serves same purpose as previous param, but we use "columns" in one and "features" in the other. Besides, they're mutually exclusive. Can't they be merged into one param? `features=5` or `features=['col_1', 'col_2']`?
- `plot_overrides`: in matplotlib world, don't we just usually use a `kwargs` to forward  those?
- `plot_overrides`, `exclude_explanations`: careful with mutable defaults (here {}, []), I always need to check that this param is not modified in the impl.
- `include_explanations`, `exclude_explanations`: just an idea, as those explanations are predefined, what aout a single  `explanations` param accepting None (=ALL) or instances of:
```
class Explanations():
     def __init__(self, leaderboard=None, ....):
            # all params default to None
            # if all set to None -> include all
            # otherwise, includes only the ones with value True, or excludes the ones with value False.

#usage
explanations=Explanations(leaderboard=True, model_correlation_heatmap=True)
explanations=Explanations(model_correlation_heatmap=False)
```
Benefits: discoverability+autocompletion for the possible explanations, include+exclude in one param.

Reply:
same for `explain_row`

Reply:
> * `columns_of_interest`: isn't it a bit verbose? why not just `columns`?
> * `top_n_features`: serves same purpose as previous param, but we use "columns" in one and "features" in the other. Besides, they're mutually exclusive. Can't they be merged into one param? `features=5` or `features=['col_1', 'col_2']`?

we should probably keep the names same in both python and R, right? I can change the behaviour in both. What do you think @ledell?


> * `plot_overrides`: in matplotlib world, don't we just usually use a `kwargs` to forward  those?

this used to be called user overrides, it is used to change parameters such as how many features to plot in one particular type of plot so it's not really the thing that I would use `kwargs` (e.g., line type, color etc)

> * `plot_overrides`, `exclude_explanations`: careful with mutable defaults (here {}, []), I always need to check that this param is not modified in the impl.
 
yep, both are read-only

> * `include_explanations`, `exclude_explanations`: just an idea, as those explanations are predefined, what about a single  `explanations` param accepting None (=ALL) or instances of:
> 
> ```
> class Explanations():
>      def __init__(self, leaderboard=None, ....):
>             # all params default to None
>             # if all set to None -> include all
>             # otherwise, includes only the ones with value True, or excludes the ones with value False.
> 
> #usage
> explanations=Explanations(leaderboard=True, model_correlation_heatmap=True)
> explanations=Explanations(model_correlation_heatmap=False)
> ```
> 
> Benefits: discoverability+autocompletion for the possible explanations, include+exclude in one param.

I like this idea, not sure if I can or should do similar thing in R. Also I'm not sure if we want the method/function arguments differ that much between R and python. Any thoughts @ledell ?

Reply:
@ledell 
- Changed the `columns_of_interest` and `top_n_features` to one parameter `features` with behaviour as Seb described above in both Python and R.

- In python changed the implementation of explanation to use Seb's idea. Usage:
```
import h2o
from h2o.explain import Explanations, RowExplantations

aml.explain(test, explanations=Explanations(leaderboard=False))
aml.explain_row(test, 1, explanations=RowExplanations(ice=True))
```

In R, it stays the same as before (include_explanations/exclude_explanations). I can easily revert this commit or change somehow R's behaviour.

Reply:
Let's change the name `features` to `columns` for consistency with other H2O functions.  We can keep the support for integers (to trigger top N) if that seems better, but I am also not opposed to having an extra arg for the latter.

Reply:
@tomasfryda @sebhrusen Regarding changing the format of the explanation args in Python and converting into a single `explanations` parameter -- it sounds like the main benefits are:

- remove a parameter
- autocompletion

It's a handy way of storing that information, but I also see some drawbacks:

- Inconsistency between R and Python (adds complexity / duplicates documentation).
- User has two extra imports
- This is not an established way of specifying params in H2O Python module (it could be in the future, but we don't have anything like that that we know of, so far, so it might seem a bit too unique to some folks...?)
- It requires bout 2x the amount of typing because now you have to put True or False in addition to the name.

I honestly don't love the whole `include_*` `exclude_*` thing, but it's got some "existing conventions" weight to it.

Reply:
@ledell seems like you'd prefer the include/exclude thing. I dropped the commit for now but I'm keeping the patch handy in case we would like to use the `Explanation` object after all...